text/2017-10-06-api-layers.md

```diff
@@ -0,0 +1,182 @@
+
+When you want to embed a garbage collector in another systems library, e.g. logger, you first define
+its own `Global` and `Participant` structs, and embed `epoch::Global` and `epoch::Participant` in
+corresponding structs, as follows:
+
+```rust
+pub mod Logger {
+
+pub struct Global {
+    epoch: epoch::Global,
+    ... // more global data
+}
+
+pub struct Participant {
+    epoch: epoch::Participant,
+    ... // more local data
+}
+
+...
+
+}
+```
+
+You can easily finish the internal logger API by implementing the methods of `Global` and
+`Participant` using `epoch`. You can build the `Logger` API and the default logger API in the same
+way as below.
+
+For a full example, see [this repository](https://github.com/jeehoonkang/handle-example-rs).
+
+
+## `Collector` API
+
+The public struct `Collector` is a general-purpose garbage collector with its own global epoch, and
+the public struct `Handle` is an abstraction of a participant in a collector.  `Collector` is a
+counted reference to a `Global`, and `Handle` is a tuple of a counted reference to a `Global` and a
+`Participant`:
+
+```rust
+struct Collector(Arc<Global>);
+
+struct Handle {
+    global: Arc<Global>,
+    local: Local,
+}
+```
+
+You can create a new collector by creating a new `Global` data, and add a handle to a collector by
+sharing its `Global` data.  Using a `Handle`, you can create a `Scope` and pass it to a given
+function:
+
+```rust
+impl Collector {
+    pub fn new() -> Self;
+    pub fn handle(&self) -> Handle;
+}
+
+impl Handle {
+    pub fn pin<F, R>(&self, f: F) 
+    where F: for<'scope> FnOnce(Scope<'scope>) -> R {
+        self.participant.pin(&self.global, f)
+    }
+
+    ... // other methods
+}
+```
+
+Note that `Handle::pin()`, on contrary to `Participant::pin()`, does not require a reference to a
+`Global`, because `Handle` already has a reference.
+
+See [the
+implementation](https://github.com/jeehoonkang/crossbeam-epoch/blob/handle/src/collector.rs) for
+more details.
+
+
+## The default garbage collector API
+
+There will be the default collector and per-thread handle to the default collector:
+
+```rust
+lazy_static! { pub GLOBAL: Global = Global::new(); }
+thread_local! { pub PARTICIPANT: Participant = Participant::new(&GLOBAL); }
+
+pub fn pin<F, R>(f: F) 
+where F: for<'scope> FnOnce(Scope<'scope>) -> R {
+    PARTICIPANT.with(|participant| { participant.pin(&GLOBAL, f) })
+}
+```
+
+See [the implementation](https://github.com/jeehoonkang/crossbeam-epoch/blob/handle/src/default.rs)
+for more details.
+
+
+
+# Alternatives
+
+`Scope<'scope>` vs. `&'scope Scope`: Currently the type of scope is `Scope<'scope>`. It is possible
+to use more traditional reference type `&'scope Scope`, but it may incur some runtime cost.
+
+`Arc<Global>` vs. `&Global`: Currently `Collector` is implemented as `Arc<Global>`, which incurs
+runtime overhead of reference counting.  Using `&Global` will eliminate the runtime cost, but it
```

Comment:
Is this a problem in practice? As far as I can see Arc::clone(..) happens when creating Collectors and Handles. The first is very likely a non issue. The second I'm not sure, but I guess creating Handles won't be in people hot path anyway.

Reply:
Thank you for the comment!  I slightly edited the text, articulating the fact that we chose `Arc<Global>` over `&Global` because of exactly what you said.